File: 2001~2500/2482. Difference Between Ones and Zeros in Row and Column/main.py
"""
Source: https://rb.gy/n86qa
Date: 2023/8/4
Skill:
Ref:
Runtime: 1414 ms, faster than 92.35%
Memory Usage: 51.81 MB, less than 58.20%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def onesMinusZeros(self, grid: List[List[int]]) -> List[List[int]]:
        m, n = len(grid), len(grid[0])
        diff = [[0 for _ in range(n)] for _ in range(m)]
        one_row, one_col = [0 for _ in range(m)], [0 for _ in range(n)]
        for i in range(m):
            cnt = 0
            for j in range(n):
                cnt += grid[i][j]
            one_row[i] = cnt
        for i in range(n):
            cnt = 0
            for j in range(m):
                cnt += grid[j][i]
            one_col[i] = cnt

        for i in range(m):
            for j in range(n):
                diff[i][j] = 2 * one_row[i] - n + 2 * one_col[j] - m

        return diff


if __name__ == "__main__":
    s = Solution()

File: 2001~2500/2482. Difference Between Ones and Zeros in Row and Column/main.cpp
/**
 * Source: https://rb.gy/n86qa
 * Date: 2023/12/14
 * Skill:
 * Ref:
 * Runtime: 185 ms, faster than 82.16% of C++ online submissions
 * Memory Usage: 116.77 MB, less than 85.57% of C++ online submissions
 * Time complexity:
 * Space complexity:
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;
#define SZ(X) ((int)(X).size())
const int MX = 1e5+1;


class Solution {
public:
    vector<vector<int>> onesMinusZeros(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> res(m, vector<int>(n, 0));
        int rows[MX] = {0}, cols[MX] = {0};
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                rows[i] += grid[i][j] == 1 ? 1 : -1;
                cols[j] += grid[i][j] == 1 ? 1 : -1;
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                res[i][j] = rows[i] + cols[j];
            }
        }
        return res;
    }
};